ord_schema/dataset_templating.py

```diff
@@ -0,0 +1,139 @@
+spreadsheet are defined, namely that placeholder values in the pbtxt begin and
+end with a "$" (dollar sign) and that these match a unique column header in the
+spreadsheet file.
+
+Example usage:
+
+* For normal dataset generation from a template reaction:
+  $ python dataset_templating.py --input_template=my_reaction.pbtxt
+      --input_spreadsheet=my_experiments.csv --output=my_dataset.pbtxt
+"""
+
+import os
+import re
+import pandas as pd
+
+from absl import app
+from absl import flags
+from absl import logging
+
+from google.protobuf import text_format
+
+from ord_schema import validations
+from ord_schema import message_helpers
+from ord_schema.proto import dataset_pb2
+from ord_schema.proto import reaction_pb2
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string('input_template', None,
+                    'Path to a Reaction pbtxt file defining a template.')
+flags.DEFINE_string(
+    'input_spreadsheet', None,
+    'Path to a spreadsheet file (with a header row) defining '
+    'values to replace placeholders in the input_template.')
+flags.DEFINE_string('output', None, 'Filename for output Dataset.')
+flags.DEFINE_boolean('validate', True, 'If True, validate Reaction protos.')
+
+
+def read_spreadsheet(file_name):
+    """Reads a spreadsheet file."""
+    _, suffix = os.path.splitext(file_name)
+    if suffix in ['xls', 'xlsx']:
+        return pd.read_excel(file_name, dtype=str)
+    return pd.read_csv(file_name, dtype=str)
+
+
+def _replace(string, substitutions):
+    pattern = re.compile('|'.join(map(re.escape, substitutions.keys())))
+    return pattern.sub(lambda match: substitutions[match.group(0)], string)
+
+
+def generate_dataset(template_string, df, validate=True):
+    """Generates a Dataset from a template reaction formatted as text according
+    to entries in a pd.Dataframe.
+
+    Inputs:
+        template_string: The contents of a Reaction pbtxt where placeholder
+            values to be replaced are defined between dollar sign. For example,
+            a SMILES identifier value could be "$product_smiles$". Spaces
+            are not allowed.
+        df: Pandas Dataframe where each row corresponds to one reaction and
+            column names match placeholders in the template_string.
+        validate: Optional Boolean controlling whether Reaction messages should
+            be validated as they are defined. Defaults to True.
+
+    Returns:
+        A Dataset message.
+
+    Raises:
+        ValueError: If there is no match for a placeholder string in df.
+        ValueError: If validate is True and there are validation errors when
+            validating an enumerated Reaction message.
+
+    """
+    placeholders = set(re.findall(r'\$\w+\$', template_string))
+    for placeholder in placeholders:
+        if placeholder not in df.columns:
+            # Allow "$my_placeholder$" to match "my_placeholder" in df.
+            if placeholder[1:-1] not in df.columns:
```

Comment:
If the current placeholder matched r'\$\w+\$', then isn't "$my_placeholder" already excluded?